paginator.go

```diff
@@ -3,177 +3,271 @@ package paging
+
+	Limit   int64       `json:"limit"`
+	NextURI null.String `json:"next"`
 }
 
 // -----------------------------------------------------------------------------
-// Page
+// Paginator with cursor
 // -----------------------------------------------------------------------------
 
-// Page contains pagination information.
-type Page struct {
-	paginator *Paginator
+// CursorPaginator is the paginator with cursor pagination system.
+type CursorPaginator struct {
+	*paginator
+	Cursor      interface{} `json:"-"`
+	PreviousURI null.String `json:"-"`
+}
 
-	Limit       int64       `json:"limit"`
-	Offset      int64       `json:"offset"`
-	Count       int64       `json:"total_count"`
-	PreviousURI null.String `json:"previous"`
-	NextURI     null.String `json:"next"`
+// NewCursorPaginator returns a new CursorPaginator instance.
+func NewCursorPaginator(store Store, request *http.Request, options *Options) (*CursorPaginator, error) {
+	if options == nil {
+		options = NewOptions()
+	}
+
+	paginator := &CursorPaginator{
+		&paginator{
+			Store:   store,
+			Options: options,
+			Request: request,
+			Limit:   GetLimitFromRequest(request, options),
+		},
+		GetCursorFromRequest(request, options),
+		null.NewString("", false),
+	}
+
+	if options.CursorOptions.Mode == DateModeCursor {
+		// time in cursor is standard timestamp (second)
+		paginator.Cursor = time.Unix(0, GetCursorFromRequest(request, options)*1000000000)
+	}
+
+	return paginator, nil
 }
 
-// Previous returns the previous page.
-func (p *Page) Previous() (*Page, error) {
-	if !p.paginator.HasPrevious() {
-		return nil, errors.New("No previous page")
+// Page searches and returns the items
+func (p *CursorPaginator) Page() error {
+	err := p.Store.PaginateCursor(
+		p.Limit,
+		p.Cursor,
+		p.Options.CursorOptions.DBName,
+		p.Options.CursorOptions.Reverse)
+	if err != nil {
+		return err
 	}
 
-	paginator := *p.paginator
-	paginator.Offset = p.Offset - p.Limit
+	p.PreviousURI = p.MakePreviousURI()
+	p.NextURI = p.MakeNextURI()
 
-	return paginator.Page()
+	return nil
 }
 
-// Next returns the next page.
-func (p *Page) Next() (*Page, error) {
-	if !p.paginator.HasNext() {
+// Previous is not available on cursor system
+func (p *CursorPaginator) Previous() (Paginator, error) {
+	return nil, errors.New("No previous page")
+}
+
+// Next returns next items
+func (p *CursorPaginator) Next() (Paginator, error) {
+	if !p.HasNext() {
 		return nil, errors.New("No next page")
 	}
 
-	paginator := *p.paginator
-	paginator.Offset = p.Offset + p.Limit
+	np := *p
+	np.Cursor = Last(p.Store.GetItems(), np.Options.CursorOptions.StructName)
+	err := np.Store.PaginateCursor(
+		np.Limit,
+		np.Cursor,
+		np.Options.CursorOptions.DBName,
+		np.Options.CursorOptions.Reverse)
+	if err != nil {
+		return nil, err
+	}
+
+	np.NextURI = p.MakeNextURI()
+
+	return &np, nil
+}
+
+// HasPrevious returns false, previous page is not available on cursor system
+func (p *CursorPaginator) HasPrevious() bool {
```

Comment:
You don't need pointer receiver here. `func (CursorPaginator) HasPrevious() bool` is enough. The same for `HasNext()` and `MakePreviousURI()`.